File: js/src/lib/args-parser-queries.js
/**
 * Query and control option parsing for the start-command wrapper.
 *
 * These options never launch a new command from scratch: they address an
 * already tracked execution by UUID or session name. They are kept in their own
 * module so `args-parser.js` stays focused on the launch options.
 *
 * Query modes:   --status, --list [--running]
 * Control modes: --upload-log, --stop, --terminate, --attach [--read-only],
 *                --resume [-- <command>], --resume-all, --cleanup
 */

/**
 * Valid output formats for --status/--list and the machine-readable
 * --resume/--resume-all summaries.
 */
const VALID_OUTPUT_FORMATS = ['links-notation', 'json', 'text'];

/**
 * Default values for every query/control option.
 * @returns {object} Fresh defaults (never share the same object between parses)
 */
function createQueryOptionDefaults() {
  return {
    status: null, // UUID/session name to show status for
    list: false, // List all tracked execution records
    running: false, // Restrict --list to executions that are still running
    uploadLog: null, // UUID/session name whose stored log should be uploaded
    outputFormat: null, // Output format (links-notation, json, text)
    stop: null, // UUID/session name to stop gracefully
    terminate: null, // UUID/session name to terminate immediately
    attach: null, // UUID/session name whose terminal should be attached
    readOnly: false, // Follow --attach output without forwarding stdin
    resume: null, // UUID/session name to resume (optionally with a new command)
    resumeAll: false, // Re-attach/resume every execution still marked running
    cleanup: false, // Clean up stale "executing" records
    cleanupDryRun: false, // Show what would be cleaned without cleaning
  };
}

/**
 * Options that take a `<uuid-or-session-name>` argument, mapped to the wrapper
 * option they populate.
 */
const IDENTIFIER_OPTIONS = {
  '--status': 'status',
  '--upload-log': 'uploadLog',
  '--stop': 'stop',
  '--terminate': 'terminate',
  '--attach': 'attach',
  '--resume': 'resume',
};

/**
 * Boolean flags with no argument.
 */
const FLAG_OPTIONS = {
  '--list': (options) => {
    options.list = true;
  },
  '--running': (options) => {
    options.running = true;
  },
  '--read-only': (options) => {
    options.readOnly = true;
  },
  '--resume-all': (options) => {
    options.resumeAll = true;
  },
  '--cleanup': (options) => {
    options.cleanup = true;
  },
  '--cleanup-dry-run': (options) => {
    options.cleanup = true;
    options.cleanupDryRun = true;
  },
};

function hasValue(value) {
  return value !== null && value !== undefined;
}

/**
 * Parse a single query/control option.
 * @param {string[]} args - Arguments array
 * @param {number} index - Current index
 * @param {object} options - Options object to populate
 * @returns {number} Number of arguments consumed (0 if not recognized)
 */
function parseQueryOption(args, index, options) {
  const arg = args[index];

  const flag = FLAG_OPTIONS[arg];
  if (flag) {
    flag(options);
    return 1;
  }

  const identifierKey = IDENTIFIER_OPTIONS[arg];
  if (identifierKey) {
    if (index + 1 < args.length && !args[index + 1].startsWith('-')) {
      options[identifierKey] = args[index + 1];
      return 2;
    }
    throw new Error(`Option ${arg} requires a UUID or session name argument`);
  }

  for (const [name, key] of Object.entries(IDENTIFIER_OPTIONS)) {
    const prefix = `${name}=`;
    if (arg.startsWith(prefix)) {
      const value = arg.slice(prefix.length);
      if (!value) {
        throw new Error(
          `Option ${name} requires a UUID or session name argument`
        );
      }
      options[key] = value;
      return 1;
    }
  }

  // --output-format <format>
  if (arg === '--output-format') {
    if (index + 1 < args.length && !args[index + 1].startsWith('-')) {
      options.outputFormat = args[index + 1].toLowerCase();
      return 2;
    }
    throw new Error(`Option ${arg} requires a format argument`);
  }

  // --output-format=<value>
  if (arg.startsWith('--output-format=')) {
    options.outputFormat = arg.split('=')[1].toLowerCase();
    return 1;
  }

  return 0;
}

/**
 * Modes that address one stored execution and therefore cannot be combined.
 * @param {object} options - Parsed wrapper options
 * @returns {string[]} Names of the modes that are active
 */
function activeQueryModes(options) {
  const modes = [];
  if (hasValue(options.status)) {
    modes.push('--status');
  }
  if (options.list) {
    modes.push('--list');
  }
  if (hasValue(options.uploadLog)) {
    modes.push('--upload-log');
  }
  if (hasValue(options.stop)) {
    modes.push('--stop');
  }
  if (hasValue(options.terminate)) {
    modes.push('--terminate');
  }
  if (hasValue(options.attach)) {
    modes.push('--attach');
  }
  if (hasValue(options.resume)) {
    modes.push('--resume');
  }
  if (options.resumeAll) {
    modes.push('--resume-all');
  }
  if (options.cleanup) {
    modes.push('--cleanup');
  }
  return modes;
}

/**
 * Formats are only meaningful for modes that print a machine-readable report.
 */
const OUTPUT_FORMAT_MODES = ['--status', '--list', '--resume', '--resume-all'];

/**
 * Validate query/control options.
 * @param {object} options - Parsed wrapper options
 * @throws {Error} When the combination is not supported
 */
function validateQueryOptions(options) {
  if (hasValue(options.outputFormat)) {
    if (!VALID_OUTPUT_FORMATS.includes(options.outputFormat)) {
      throw new Error(
        `Invalid output format: "${options.outputFormat}". Valid options are: ${VALID_OUTPUT_FORMATS.join(', ')}`
      );
    }
  }

  const modes = activeQueryModes(options);

  if (modes.length > 1) {
    throw new Error(
      'Cannot combine --status, --list, --upload-log, --stop, --terminate, --attach, --resume, --resume-all, or --cleanup in the same invocation'
    );
  }

  if (
    hasValue(options.outputFormat) &&
    !modes.some((mode) => OUTPUT_FORMAT_MODES.includes(mode))
  ) {
    throw new Error(
      '--output-format option is only valid with --status, --list, --resume, or --resume-all'
    );
  }

  if (options.running && !options.list) {
    throw new Error('--running option is only valid with --list');
  }

  if (options.readOnly && !hasValue(options.attach)) {
    throw new Error('--read-only option is only valid with --attach');
  }
}

module.exports = {
  VALID_OUTPUT_FORMATS,
  activeQueryModes,
  createQueryOptionDefaults,
  parseQueryOption,
  validateQueryOptions,
};

File: rust/src/lib/args_parser_queries.rs
//! Query/control option parsing for the start-command wrapper.
//!
//! These options address an execution that already exists instead of starting
//! a new one: `--status`, `--list [--running]`, `--upload-log`, `--stop`,
//! `--terminate`, `--attach [--read-only]`, `--resume [-- <command>]`,
//! `--resume-all` and `--cleanup`. They are parsed here so `args_parser` stays
//! focused on the options that shape a new execution (issue #162).

use crate::args_parser::{WrapperOptions, VALID_OUTPUT_FORMATS};

/// Setter applied to the parsed options when an identifier option matches.
type IdentifierSetter = fn(&mut WrapperOptions, String);

/// Setter applied to the parsed options when a flag option matches.
type FlagSetter = fn(&mut WrapperOptions);

/// Options that take a UUID or session name argument.
const IDENTIFIER_OPTIONS: [(&str, IdentifierSetter); 6] = [
    ("--status", |o, v| o.status = Some(v)),
    ("--upload-log", |o, v| o.upload_log = Some(v)),
    ("--stop", |o, v| o.stop = Some(v)),
    ("--terminate", |o, v| o.terminate = Some(v)),
    ("--attach", |o, v| o.attach = Some(v)),
    ("--resume", |o, v| o.resume = Some(v)),
];

/// Options that are plain flags.
const FLAG_OPTIONS: [(&str, FlagSetter); 6] = [
    ("--list", |o| o.list = true),
    ("--running", |o| o.running = true),
    ("--read-only", |o| o.read_only = true),
    ("--resume-all", |o| o.resume_all = true),
    ("--cleanup", |o| o.cleanup = true),
    ("--cleanup-dry-run", |o| {
        o.cleanup = true;
        o.cleanup_dry_run = true;
    }),
];

/// Query modes that may be combined with `--output-format`.
const OUTPUT_FORMAT_MODES: [&str; 4] = ["--status", "--list", "--resume", "--resume-all"];

/// Parse one query/control option.
///
/// Returns the number of arguments consumed, or `Ok(0)` when `args[index]` is
/// not a query/control option.
pub fn parse_query_option(
    args: &[String],
    index: usize,
    options: &mut WrapperOptions,
) -> Result<usize, String> {
    let arg = &args[index];

    for (name, apply) in FLAG_OPTIONS {
        if arg == name {
            apply(options);
            return Ok(1);
        }
    }

    for (name, apply) in IDENTIFIER_OPTIONS {
        if arg == name {
            if index + 1 < args.len() && !args[index + 1].starts_with('-') {
                apply(options, args[index + 1].clone());
                return Ok(2);
            }
            return Err(format!(
                "Option {} requires a UUID or session name argument",
                arg
            ));
        }

        let prefix = format!("{}=", name);
        if let Some(value) = arg.strip_prefix(&prefix) {
            if value.is_empty() {
                return Err(format!(
                    "Option {} requires a UUID or session name argument",
                    name
                ));
            }
            apply(options, value.to_string());
            return Ok(1);
        }
    }

    if arg == "--output-format" {
        if index + 1 < args.len() && !args[index + 1].starts_with('-') {
            options.output_format = Some(args[index + 1].to_lowercase());
            return Ok(2);
        }
        return Err(format!("Option {} requires a format argument", arg));
    }

    if arg.starts_with("--output-format=") {
        options.output_format = Some(arg.split('=').nth(1).unwrap_or("").to_lowercase());
        return Ok(1);
    }

    Ok(0)
}

/// List the query/control modes selected by these options.
pub fn active_query_modes(options: &WrapperOptions) -> Vec<&'static str> {
    let mut modes = Vec::new();
    if options.status.is_some() {
        modes.push("--status");
    }
    if options.list {
        modes.push("--list");
    }
    if options.upload_log.is_some() {
        modes.push("--upload-log");
    }
    if options.stop.is_some() {
        modes.push("--stop");
    }
    if options.terminate.is_some() {
        modes.push("--terminate");
    }
    if options.attach.is_some() {
        modes.push("--attach");
    }
    if options.resume.is_some() {
        modes.push("--resume");
    }
    if options.resume_all {
        modes.push("--resume-all");
    }
    if options.cleanup {
        modes.push("--cleanup");
    }
    modes
}

/// True when any query/control mode is active.
pub fn has_query_mode(options: &WrapperOptions) -> bool {
    !active_query_modes(options).is_empty()
}

/// Validate query/control option combinations.
pub fn validate_query_options(options: &WrapperOptions) -> Result<(), String> {
    if let Some(ref format) = options.output_format {
        if !VALID_OUTPUT_FORMATS.contains(&format.as_str()) {
            return Err(format!(
                "Invalid output format: \"{}\". Valid options are: {}",
                format,
                VALID_OUTPUT_FORMATS.join(", ")
            ));
        }
    }

    let modes = active_query_modes(options);

    if modes.len() > 1 {
        return Err(
            "Cannot combine --status, --list, --upload-log, --stop, --terminate, --attach, --resume, --resume-all, or --cleanup in the same invocation"
                .to_string(),
        );
    }

    if options.output_format.is_some()
        && !modes.iter().any(|mode| OUTPUT_FORMAT_MODES.contains(mode))
    {
        return Err(
            "--output-format option is only valid with --status, --list, --resume, or --resume-all"
                .to_string(),
        );
    }

    if options.running && !options.list {
        return Err("--running option is only valid with --list".to_string());
    }

    if options.read_only && options.attach.is_none() {
        return Err("--read-only option is only valid with --attach".to_string());
    }

    Ok(())
}
